queryable-spec.ts

```diff
@@ -0,0 +1,390 @@
+  type: 'bindings' | 'quads' | 'boolean';
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<MetadataOrderType>>;
+}
+
+interface QueryableResultBindings extends BaseQueryableResult<RDF.Variable> {
+  type: 'bindings';
+  stream(): Stream<Bindings>;
+  variables: RDF.Variable[];
+}
+    
+interface QueryableResultQuads extends BaseQueryableResult<TermName> {
+  type: 'quads';
+  stream(): Stream<RDF.Quad>;
+}
+
+interface QueryableResultBoolean extends BaseQueryableResult<any> {
+  type: 'boolean';
+  value(): Promise<boolean>;
+}
+
+type QueryableResult = QueryableResultBindings | QueryableResultQuads | QueryableResultBoolean;/*
+ * Context objects provide a way to pass additional bits information to
+ * implementors, such as but not limited to:
+ * - data sources
+ * - base IRI for IRI resolution
+ * - timestamp for expression evaluation
+ * - query language
+ * - ...
+ */
+
+// SourceType can be anything the query engine defines
+// TODO: we may consider defining some standards, like 'string', RDF.Source, ...
+interface QueryableContext<SourceType> {
+  sources: [SourceType, ...SourceType[]];
+  queryTimestamp?: Date; // Required for certain SPARQL operations such as NOW().
+  [key: string]: any;
+}
+    
+interface QueryableStringContext<SourceType> extends QueryableContext<SourceType> {
+  queryFormat?: QueryableFormat; // defaults to { language: 'SPARQL', version: '1.1', extensions: [] }
+  baseIRI?: string; // Required for parsing SPARQL queries
+};
+
+interface QueryableAlgebraContext<SourceType> extends QueryableContext<SourceType> {};
+    
+interface QueryableFormat {
+  language: string; // Like 'SPARQL'
+  version: string; // Like '1.1'
+  extensions: string[]; // TODO: leave the syntax of these extensions open for now?
+}
+
+/* 
+ * Generic query interfaces. These allow engines to return any type of result
+ * object for any type of query, supporting the kind of flexibility required
+ * by engines such as Comunica.
+ */
+
+interface Queryable<SourceType> {
+  query(query: string, context?: QueryableStringContext<SourceType>): Promise<QueryableResult>;
+}
+    
+interface QueryableAlgebra<SourceType> {
+  query(query: Algebra.Operation, context?: QueryableAlgebraContext<SourceType>): Promise<QueryableResult>;
+}
+
+/*
+ * SPARQL-constrainted query interfaces. These interfaces guarantee that result
+ * objects are of the expected type as defined by the SPARQL spec.
+ */
+
+interface QueryableSparql<SourceType> {
+  boolean?(query: string, context?: QueryableContext<SourceType>): Promise<QueryableResultBoolean>;
+  bindings?(query: string, context?: QueryableContext<SourceType>): Promise<QueryableResultBindings>;
+  quads?(query: string, context?: QueryableContext<SourceType>): Promise<QueryableResultQuads>;
+}
+
+interface QueryableAlgebraSparql<SourceType> {
+  boolean?(query: Algebra.Ask, context?: QueryableAlgebraContext<SourceType>): Promise<QueryableResultBoolean>;
```

Comment:
Since defining this algebra in RDF/JS might require quite a bit of effort, my suggestion is to leave this algebra unspecified for now, and refer to https://github.com/joachimvh/SPARQLAlgebra.js as one possible format.

Reply:
Good point!